### PR TITLE
Update Sql Parser and Column Name

### PR DESCRIPTION
**Pull request by ccfromms**
## Description

- We find that TSql140Parser does not support function like string_agg within group(order by), so we try to update DacFx and change TSql140Parser to TSql150Parser. 
- We notice that if we define sql table column name like TableName, it will conflict with TableName in Table, so we try to convert such column name to another one.

## Testing
We add a stored procedure which use the function string_agg within group(order by), and test whether this function can be identified.

### CODE REVIEW

File: tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj [hunk 1]
```diff
@@ -1,8 +1,12 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>$(SupportedFrameworks)</TargetFrameworks>
+    <TargetFrameworks>net462</TargetFrameworks>
     <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
+    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
   </PropertyGroup>
+  <ItemGroup>
+    <None Remove="Sql\sprc_string_agg.sql" />
```
Comment (wsugarman) [inline tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj:8]: Why is this removed?

File: tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj [hunk 2]
```diff
@@ -1,8 +1,12 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>$(SupportedFrameworks)</TargetFrameworks>
+    <TargetFrameworks>net462</TargetFrameworks>
     <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
+    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
```
Comment (wsugarman) [inline tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj:5]: This should be inherited via `Directory.Build.props`

File: tools/Microsoft.Health.Tools.Sql.Tasks/Helpers/SqlScriptWriter.cs [hunk 3]
```diff
@@ -22,7 +22,7 @@ public SqlScriptWriter(string path)
         Directory.CreateDirectory(Path.GetDirectoryName(path));
 
         _writer = File.CreateText(path);
-        _sqlScriptGenerator = new Sql140ScriptGenerator();
+        _sqlScriptGenerator = new Sql160ScriptGenerator();
```
Comment (wsugarman) [inline tools/Microsoft.Health.Tools.Sql.Tasks/Helpers/SqlScriptWriter.cs:25]: I take it this refers to SQL Server 16 which is 2022 and I think is in preview. Could we use the current production version 15 instead? Is there a `Sql150ScriptGenerator`?

File: tools/Microsoft.Health.Tools.Sql.Tasks/Microsoft.Health.Tools.Sql.Tasks.csproj [hunk 4]
```diff
@@ -1,14 +1,14 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>net462</TargetFramework>
```
Comment (wsugarman) [inline tools/Microsoft.Health.Tools.Sql.Tasks/Microsoft.Health.Tools.Sql.Tasks.csproj:4]: Again, we don't want to use .NET Framework, right? Alternatively we could multi-target if we needed to, but I see this package has three targets: https://www.nuget.org/packages/Microsoft.SqlServer.DacFx/160.6296.0

**Review by wsugarman — CHANGES_REQUESTED** [review]

File: tools/Microsoft.Health.Tools.Sql.Tasks/Microsoft.Health.Tools.Sql.Tasks.csproj (hunk 4 above)
Comment (wsugarman) [reply tools/Microsoft.Health.Tools.Sql.Tasks/Microsoft.Health.Tools.Sql.Tasks.csproj:4]: Also, what's the difference between this package and the old one?

File: Directory.Packages.props [hunk 5]
```diff
@@ -1,5 +1,4 @@
 <Project>
-
   <ItemGroup>
```
[the remark this answers is not in the document]
Comment (ccfromms) [reply Directory.Packages.props:2]: Sorry, I will modify this file and correspond with main branch.

File: tools/Microsoft.Health.Extensions.BuildTimeCodeGenerator/Sql/SqlVisitor.cs [hunk 6]
```diff
@@ -22,6 +22,8 @@ namespace Microsoft.Health.Extensions.BuildTimeCodeGenerator.Sql;
 /// </summary>
 public abstract class SqlVisitor : TSqlFragmentVisitor
 {
+    private static readonly ISet<string> Keywords = new HashSet<string>() { "TableName" };
```
[the remark this answers is not in the document]
Comment (ccfromms) [reply tools/Microsoft.Health.Extensions.BuildTimeCodeGenerator/Sql/SqlVisitor.cs:25]: Ok, I will replace ISet with HashSet.

[the remark this answers is not in the document]
Comment (ccfromms) [reply tools/Microsoft.Health.Extensions.BuildTimeCodeGenerator/Sql/SqlVisitor.cs:25]: TableName is not reserved, but in [Table](https://github.com/ccfromms/healthcare-shared-components/blob/main/src/Microsoft.Health.SqlServer/Features/Schema/Model/Table.cs), we have already defined a property named TableName. If I want to set a column named TableName in a certain sql table, it will throw an exception since TableName has been used in class Table.

File: tools/Microsoft.Health.Extensions.BuildTimeCodeGenerator/Sql/SqlVisitor.cs [hunk 7]
```diff
@@ -266,7 +273,7 @@ protected static MemberDeclarationSyntax CreatePropertyForTableColumn(ColumnDefi
 
         return FieldDeclaration(
                 VariableDeclaration(typeName)
-                    .AddVariables(VariableDeclarator($"{column.ColumnIdentifier.Value}")
+                    .AddVariables(VariableDeclarator(GetColunmName(column.ColumnIdentifier.Value))
```
[the remark this answers is not in the document]
Comment (ccfromms) [reply tools/Microsoft.Health.Extensions.BuildTimeCodeGenerator/Sql/SqlVisitor.cs:276]: In FHIR server, we try to define a table which has a column named TableName, and TableName is conflict with the property in class [Table](https://github.com/ccfromms/healthcare-shared-components/blob/main/src/Microsoft.Health.SqlServer/Features/Schema/Model/Table.cs). Therefor we think that it is more appropriate to transform TableName here.

File: tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj (hunk 2 above)
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj:5]: Sorry, I will remove this line.

File: tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj (hunk 1 above)
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj:8]: Sorry, I will check and fix this lines.

File: tools/Microsoft.Health.Tools.Sql.Tasks/Helpers/SqlScriptWriter.cs (hunk 3 above)
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks/Helpers/SqlScriptWriter.cs:25]: Ok, I will try and verify Sql150ScriptGenerator.

File: tools/Microsoft.Health.Tools.Sql.Tasks/Helpers/SqlScriptParser.cs [hunk 8]
```diff
@@ -31,7 +31,7 @@ public static TSqlFragment ParseSqlFile(string sqlFile, TaskLoggingHelper taskLo
         using (var stream = File.OpenRead(sqlFile))
         using (var reader = new StreamReader(stream))
         {
-            var parser = new TSql140Parser(true);
+            var parser = new TSql160Parser(true);
```
[the remark this answers is not in the document]
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks/Helpers/SqlScriptParser.cs]: Sorry, I will fix it in next commit.

File: tools/Microsoft.Health.Extensions.BuildTimeCodeGenerator/Sql/SqlVisitor.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ccfromms) [reply tools/Microsoft.Health.Extensions.BuildTimeCodeGenerator/Sql/SqlVisitor.cs:276]: Ok, I will modify the name.

File: tools/Microsoft.Health.Tools.Sql.Tasks/Sql.targets [hunk 9]
```diff
@@ -28,6 +28,6 @@
             SqlScript="@(SqlScript)" />
     </Target>
 
-    <UsingTask TaskName="GenerateFullScript" AssemblyFile="netstandard2.0\Microsoft.Health.Tools.Sql.Tasks.dll" />
+    <UsingTask TaskName="GenerateFullScript" AssemblyFile="netstandard2.1\Microsoft.Health.Tools.Sql.Tasks.dll" />
```
[the remark this answers is not in the document]
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks/Sql.targets]: Sorry, I tried it again and found that MSBuild cannot support .NET Standard2.1. I will modify it in next commit.

File: tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj [hunk 10]
```diff
@@ -27,9 +28,12 @@
     <GeneratedFullScriptPath>Schema\$(LatestSchemaVersion).sql</GeneratedFullScriptPath>
   </PropertyGroup>
 
-  <Target Name="PublishSqlTasks" BeforeTargets="CoreCompile">
-    <MSBuild Projects="..\Microsoft.Health.Tools.Sql.Tasks\Microsoft.Health.Tools.Sql.Tasks.csproj" Targets="Publish" Properties="TargetFramework=netstandard2.0" />
+  <Target Name="PublishSqlTasks" BeforeTargets="CoreCompile" Condition=" '$(MSBuildRuntimeType)' == 'Core' ">
```
Comment (wsugarman) [inline tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj:31]: Alternative you could use a [`<choose>`](https://learn.microsoft.com/en-us/visualstudio/msbuild/choose-element-msbuild?view=vs-2022) element.

File: tools/Microsoft.Health.Tools.Sql.Tasks/Sql.targets [hunk 11]
```diff
@@ -28,6 +28,7 @@
             SqlScript="@(SqlScript)" />
     </Target>
 
-    <UsingTask TaskName="GenerateFullScript" AssemblyFile="netstandard2.0\Microsoft.Health.Tools.Sql.Tasks.dll" />
+    <UsingTask TaskName="GenerateFullScript" AssemblyFile="netstandard2.1\Microsoft.Health.Tools.Sql.Tasks.dll" Condition=" '$(MSBuildRuntimeType)' == 'Core' "/>
```
Comment (wsugarman) [inline tools/Microsoft.Health.Tools.Sql.Tasks/Sql.targets:31]: Probably more appropriate to check would be `'$(MSBuildRuntimeType)' == 'Full'` as I think .NET Standard 2.1 should work with .NET Core/.NET or Mono

**Review by wsugarman — APPROVED** [review]

File: tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj (hunk 10 above)
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj:31]: Ok, I will change this element.

File: tools/Microsoft.Health.Tools.Sql.Tasks/Sql.targets (hunk 11 above)
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks/Sql.targets:31]: Thanks, I will fix it now.

File: tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj (hunk 10 above)
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks.Tests/Microsoft.Health.Tools.Sql.Tasks.Tests.csproj]: The choose tag should be the same level like target as this [link ](https://stackoverflow.com/questions/58090751/the-element-when-benerath-element-choose-is-unrecognized)said, so I think it will be better to use condition here.

File: tools/Microsoft.Health.Tools.Sql.Tasks/Sql.targets (hunk 11 above)
Comment (ccfromms) [reply tools/Microsoft.Health.Tools.Sql.Tasks/Sql.targets]: I have tried it according to [MSBuildRuntimeType ](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-reserved-and-well-known-properties?view=vs-2022#:~:text=Full%20indicating%20that%20MSBuild%20is%20running%20on%20the%20desktop%20.NET%20Framework%2C%20Core%20indicating%20that%20MSBuild%20is%20running%20on%20.NET%20Core%20(for%20example%20in%20dotnet%20build)%2C%20or%20Mono%20indicating%20that%20MSBuild%20is%20running%20on%20Mono.)and found out that it will work if we use 'Full' to specify .NET Framework and use 'Core' and 'Mono' to specify .NET Core and Mono.
